src/naive_bayes/mod.rs

```diff
@@ -0,0 +1,186 @@
+
+/// Distribution used in the Naive Bayes classifier.
+pub trait NBDistribution<T: RealNumber, M: Matrix<T>> {
+    /// Fits the distribution to a NxM matrix where N is number of samples and M is number of features.
+    /// * `x` - training data
+    /// * `y` - vector with target values (classes) of length N
+    fn fit(x: &M, y: &M::RowVector) -> Self;
+
+    /// Prior of class k
+    fn prior(&self, k: T) -> T;
+
+    /// Conditional probability of feature j given class k
+    fn conditional_probability(&self, k: T, j: &M::RowVector) -> T;
+
+    /// Possible classes of the distribution
+    fn classes(&self) -> Vec<T>;
+}
+
+/// Base struct for the Naive Bayes classifier.
+pub struct BaseNaiveBayes<T: RealNumber, M: Matrix<T>, D: NBDistribution<T, M>> {
+    distribution: D,
+    _phantom_t: PhantomData<T>,
+    _phantom_m: PhantomData<M>,
+}
+
+impl<T: RealNumber, M: Matrix<T>, D: NBDistribution<T, M>> BaseNaiveBayes<T, M, D> {
+    /// Fits NB classifier to a given NBdistribution.
+    /// * `distribution` - NBDistribution of the training data
+    pub fn fit(distribution: D) -> Result<Self, Failed> {
+        Ok(Self {
+            distribution,
+            _phantom_t: PhantomData,
+            _phantom_m: PhantomData,
+        })
+    }
+
+    /// Estimates the class labels for the provided data.
+    /// * `x` - data of shape NxM where N is number of data points to estimate and M is number of features.
+    /// Returns a vector of size N with class estimates.
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        let y_classes = self.distribution.classes();
+        let (rows, _) = x.shape();
+        let predictions = (0..rows)
+            .map(|row_index| {
+                let row = x.get_row(row_index);
+                let (prediction, _probability) = y_classes
+                    .iter()
+                    .map(|class| {
+                        (
+                            class,
+                            self.distribution.conditional_probability(*class, &row)
+                                * self.distribution.prior(*class),
+                        )
+                    })
+                    .max_by(|(_, p1), (_, p2)| p1.partial_cmp(p2).unwrap())
+                    .unwrap();
+                prediction.clone()
+            })
+            .collect::<Vec<T>>();
+        let mut y_hat = M::RowVector::zeros(rows);
+        for i in 0..rows {
+            y_hat.set(i, predictions[i]);
+        }
+        Ok(y_hat)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::linalg::naive::dense_matrix::DenseMatrix;
+    struct FakeDistribution<T: RealNumber, M: Matrix<T>> {
+        y_labels: Vec<T>,
+        label_count: Vec<usize>,
+        data_size: usize,
+        x: M,
+        y: M::RowVector,
+    }
+
+    impl<T: RealNumber, M: Matrix<T>> NBDistribution<T, M> for FakeDistribution<T, M> {
+        fn fit(x: &M, y: &M::RowVector) -> Self {
+            let x = x.clone();
```

Comment:
for the MultinomialNBDistribution let's make sure we don't do excessive cloning. For example, [Multinomial distribution is parametrized by parameter _theta_](https://scikit-learn.org/stable/modules/naive_bayes.html#multinomial-naive-bayes) hence all we need to successfully execute methods _prior_, _conditional_probability_ and _classes_ are thetas, class probabilities and class values. We do not need to keep _x_ and _y_ in the struct that represents the distribution.